File: Merge Two Sorted List/Java/mergeTwoSorted-solution1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode current, head;
        // If either list is null return the other list
        if(l1 == null) return l2;
        if(l2 == null) return l1;
        // Assign current to whatever list's head is smallest. Then advance the list used
        if(l1.val <= l2.val){
            current = l1;
            l1 = l1.next;
        }else{
            current = l2;
            l2 = l2.next;
        }
        
        head = current;
        // If either list is null break from the loop
        while(l1 != null || l2 != null){
            // if either list is null the rest of the sorted list is the remained of the other list
            if(l1 == null || l2 == null){
                current.next = l1 == null ? l2 : l1;
                break;
            }
    
            // The next node is the node who's value is least between the heads of l1 and l2. Assign to next and advance the used list
            if(l1.val < l2.val){
                current.next = l1;
                l1 = l1.next;
            }else{
                current.next = l2;
                l2 = l2.next;
                }
                // Update current to be the recently added node
                current = current.next;
        }

        return head;   
    }
}

File: Merge Two Sorted List/Go/mergeTwoSorted-solution1.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	// if list1 or list2 is emptry return the other list
	if list1 == nil {
		return list2
	}
	if list2 == nil {
		return list1
	}

	// The head of our sorted list
	var head *ListNode
	// A runner node to track our place as we build the sorted list
	var runner *ListNode
	// Find the smallest value between the two heads of list1 & list2. Advance the list who's head is used
	if list1.Val > list2.Val {
		head = list2
		list2 = list2.Next
	} else {
		head = list1
		list1 = list1.Next
	}
	// Assign the runner to the head of our sorted list
	runner = head

	for runner != nil {
		// if both list are nil break since we have visited every node
		if list1 == nil && list2 == nil {
			break
		}

		// If one list is empty and the other is not add the remainder to runner and break
		if list1 == nil && list2 != nil {
			runner.Next = list2
			break
		}
		if list2 == nil && list1 != nil {
			runner.Next = list1
			break
		}

		// If both list are not nil the next node is whichever value is smaller. Advance the list that is used
		if list1 != nil && list2 != nil {
			if list1.Val < list2.Val {
				runner.Next = list1
				list1 = list1.Next
			} else {
				runner.Next = list2
				list2 = list2.Next
			}
		}

		// Advance the runner to the next node
		runner = runner.Next
	}

	// Return the head Node
	return head
}